src/types.rs

```diff
@@ -764,3 +774,190 @@ impl LateLintPass for AbsurdExtremeComparisons {
+            }),
+            _ => None,
+        }
+    } else {
+        None
+    }
+}
+
+fn node_as_const_fullint(cx: &LateContext, expr: &Expr) -> Option<FullInt> {
+    use rustc::middle::const_eval::EvalHint::ExprTypeChecked;
+    use rustc_const_eval::*;
+
+    match const_eval::eval_const_expr_partial(cx.tcx, expr, ExprTypeChecked, None) {
+        Ok(val) => {
+            if let Integral(const_int) = val {
+                Some(match const_int {
+                    I8(x) => FullInt::S(x as i64),
+                    I16(x) => FullInt::S(x as i64),
+                    I32(x) => FullInt::S(x as i64),
+                    Isize(x) => FullInt::S(match x {
+                        Is32(x_) => x_ as i64,
+                        Is64(x_) => x_
+                    }),
+                    I64(x) => FullInt::S(x),
+                    InferSigned(x) => FullInt::S(x as i64),
+                    U8(x) => FullInt::U(x as u64),
+                    U16(x) => FullInt::U(x as u64),
+                    U32(x) => FullInt::U(x as u64),
+                    Usize(x) => FullInt::U(match x {
+                        Us32(x_) => x_ as u64,
+                        Us64(x_) => x_,
+                    }),
+                    U64(x) => FullInt::U(x),
+                    Infer(x) => FullInt::U(x as u64),
+                })
+            } else {
+                None
+            }
+        },
+        Err(_) => None,
+    }
+}
+
+impl LateLintPass for InvalidUpcastComparisons {
+    fn check_expr(&mut self, cx: &LateContext, expr: &Expr) {
+        if let ExprBinary(ref cmp, ref lhs, ref rhs) = expr.node {
+            let normalized = normalize_comparison(cmp.node, lhs, rhs);
+            if normalized.is_none() { return; }
+            let (rel, normalized_lhs, normalized_rhs) = normalized.expect("Unreachable-- is none check above");
+
+            let lhs_bounds = numeric_cast_precast_bounds(cx, normalized_lhs);
+            let rhs_bounds = numeric_cast_precast_bounds(cx, normalized_rhs);
+
+            let msg = "Because of the numeric bounds prior to casting, this expression is always ";
+
+            if let Some(nlb) = lhs_bounds {
+                if let Some(norm_rhs_val) = node_as_const_fullint(cx, normalized_rhs) {
+                    if match rel {
+                        Rel::Lt => nlb.1 < norm_rhs_val,
+                        Rel::Le => nlb.1 <= norm_rhs_val,
+                    } {
+                        // Expression is always true
+                        cx.span_lint(INVALID_UPCAST_COMPARISONS,
+                                     expr.span,
+                                     &format!("{}{}.", msg, "true"));
+                    } else if match rel {
+                        Rel::Lt => nlb.0 >= norm_rhs_val,
+                        Rel::Le => nlb.0 > norm_rhs_val,
+                    } {
+                        // Expression is always false
+                        cx.span_lint(INVALID_UPCAST_COMPARISONS,
```

Comment:
Also, I forgot about that yesterday, don’t use `cx.span*` but one of [our `span_*`](https://github.com/Manishearth/rust-clippy/blob/deb0f3120d61ca948d3ef8982cc565b2547cebca/src/utils/mod.rs#L472-L524). They add a link to Clippy’s wiki.